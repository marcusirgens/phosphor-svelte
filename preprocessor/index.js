import MagicString from "magic-string"

/**
 * @returns {import(".").PreprocessorGroup}
 */
export function phosphorSvelteOptimize() {
  return {
    name: "phosphor-svelte-optimize",
    script({ attributes, filename, content }) {
      if (!filename || /node_modules/.test(filename)) return

      const re = new RegExp(
        /(?<indent>\n[ \t]+)?import\s*{(?<imports>[\s\S]*?)\s*}\s*from\s*["']phosphor-svelte["']([\s\n]*;)?/g
      )

      let output = new MagicString(content, { filename })
      for (let match of content.matchAll(re)) {
        const modules = match.groups["imports"]
          .trim()
          .replace(/,$/, "")
          .split(",")
          .map((s) => s.trim())
        const icons = modules.filter((s) => !s.startsWith("type"))
        const types = modules.filter((s) => s.startsWith("type"))

        const newImports = icons.map((icon) => {
          let matches
          if ((matches = icon.match(/(\w+)\s+as\s+(\w+)/))) {
            const module = matches[1]
            const alias = matches[2]

            return `import ${alias} from "phosphor-svelte/lib/${module}";`
          }

          return `import ${icon} from "phosphor-svelte/lib/${icon}";`
        })

        if (types.length > 0) {
          const typeImports = types.map((t) => t.slice(5, t.length)).join(", ")
          newImports.push(
            `import type { ${typeImports} } from "phosphor-svelte";`
          )
        }

        // Indent the contents to match the surrounding imports.
        // Note that the indent match, if present, starts with a newline character.
        const toInsert = new MagicString(newImports.join("\n")).indent(match.groups["indent"]?.substring(1));

        output.update(
          match.index + (!!match.groups["indent"] ? 1 : 0),
          match.index + match[0].length,
          toInsert.toString()
        )
      }

      return {
        attributes,
        code: output.toString(),
        map: output.generateMap(),
      }
    },
  }
}
