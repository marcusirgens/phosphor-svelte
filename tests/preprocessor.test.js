import { describe, expect, it } from "vitest"
import { phosphorSvelteOptimize } from "../preprocessor"
import { preprocess } from "svelte/compiler"

describe("preprocessor", () => {
  describe("phosphorSvelteOptimize", () => {
    it("should have a name", () => {
      const processor = phosphorSvelteOptimize()

      expect(processor).toHaveProperty("name", "phosphor-svelte-optimize")
    })

    it("should preprocess icon imports", async () => {
      const processor = phosphorSvelteOptimize()

      const content = `<script>
  import Airplane from "phosphor-svelte/lib/Airplane";
  import { Cube, Rectangle, Horse } from "phosphor-svelte";
</script>\n`

      const result = await preprocess(content, processor, {
        filename: "App.svelte",
      })
      expect(result).toMatchSnapshot()
    })

    it("should preprocess multiple imports", async () => {
      const processor = phosphorSvelteOptimize()

      const content = `<script>
  import { Airplane, Gear } from "phosphor-svelte";
  import { Cube, Rectangle, Horse } from "phosphor-svelte";
  import { X } from "phosphor-svelte";
</script>\n`

      const result = await preprocess(content, processor, {
        filename: "App.svelte",
      })
      expect(result).toMatchSnapshot()
    })

    it("should separate type imports", async () => {
      const processor = phosphorSvelteOptimize()

      const content = `<script lang="ts">
  import { Airplane, Gear, type IconContextProps } from "phosphor-svelte";
</script>\n`

      const result = await preprocess(content, processor, {
        filename: "App.svelte",
      })
      expect(result).toMatchSnapshot()
    })

    it("should preprocess aliased imports", () => {
      const processor = phosphorSvelteOptimize()

      const content = `<script>
  import { Airplane, Gear as IconGear } from "phosphor-svelte";
</script>\n`
      const output = `<script>
  import Airplane from "phosphor-svelte/lib/Airplane";
  import IconGear from "phosphor-svelte/lib/Gear";
</script>\n`

      const result = processor.script({ content, filename: "App.svelte" })

      expect(result.code).toBe(output)
    })

    describe("should parse", () => {
      it("single quotes", async () => {
        const processor = phosphorSvelteOptimize()
        const content = `<script>
    import { Airplane } from 'phosphor-svelte';
  </script>\n`;

        const result = await preprocess(content, processor, {
          filename: "App.svelte",
        })
        const output = `<script>
    import Airplane from "phosphor-svelte/lib/Airplane";
  </script>\n`

        expect(result.code).toBe(output)
      })

      it("multiline imports", async () => {
        const processor = phosphorSvelteOptimize()
        const content = `<script lang="ts">
    import { 
        Airplane,
        type Hourglass as HourglassAlias,
        FolderNotch,
    } from "phosphor-svelte";
    import { Ghost } from 'phosphor-svelte';
  </script>\n`;

        const result = await preprocess(content, processor, {
          filename: "App.svelte",
        })
        const output = `<script lang="ts">
    import Airplane from "phosphor-svelte/lib/Airplane";
    import FolderNotch from "phosphor-svelte/lib/FolderNotch";
    import type { Hourglass as HourglassAlias } from "phosphor-svelte";
    import Ghost from "phosphor-svelte/lib/Ghost";
  </script>\n`

        expect(result.code).toBe(output)
      })
    })
  })
})
