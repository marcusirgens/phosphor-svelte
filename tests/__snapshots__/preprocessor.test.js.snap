// Vitest Snapshot v1

exports[`preprocessor > phosphorSvelteOptimize > should preprocess icon imports 1`] = `
{
  "code": "<script>
  import Airplane from \\"phosphor-svelte/lib/Airplane\\";
  import Cube from \\"phosphor-svelte/lib/Cube\\";
  import Rectangle from \\"phosphor-svelte/lib/Rectangle\\";
  import Horse from \\"phosphor-svelte/lib/Horse\\";
</script>
",
  "dependencies": [],
  "map": SourceMap {
    "mappings": "AAAA,CAAC,MAAM,CCAP;AACA;AACA;AAAA;AAAA,gDAA2D;ADC3D,CAAC,CAAC,MAAM;",
    "names": [],
    "sources": [
      "App.svelte",
      "",
    ],
    "version": 3,
  },
  "toString": [Function],
}
`;

exports[`preprocessor > phosphorSvelteOptimize > should preprocess multiple imports 1`] = `
{
  "code": "<script>
  import Airplane from \\"phosphor-svelte/lib/Airplane\\";
  import Gear from \\"phosphor-svelte/lib/Gear\\";
  import Cube from \\"phosphor-svelte/lib/Cube\\";
  import Rectangle from \\"phosphor-svelte/lib/Rectangle\\";
  import Horse from \\"phosphor-svelte/lib/Horse\\";
  import X from \\"phosphor-svelte/lib/X\\";
</script>
",
  "dependencies": [],
  "map": SourceMap {
    "mappings": "AAAA,CAAC,MAAM,CCAP;AACA;AAAA,8CAAmD;AACnD;AAAA;AAAA,gDAA2D;AAC3D,wCAAsC;ADCtC,CAAC,CAAC,MAAM;",
    "names": [],
    "sources": [
      "App.svelte",
      "",
    ],
    "version": 3,
  },
  "toString": [Function],
}
`;

exports[`preprocessor > phosphorSvelteOptimize > should separate type imports 1`] = `
{
  "code": "<script lang=\\"ts\\">
  import Airplane from \\"phosphor-svelte/lib/Airplane\\";
  import Gear from \\"phosphor-svelte/lib/Gear\\";
  import type { IconContextProps } from \\"phosphor-svelte\\";
</script>
",
  "dependencies": [],
  "map": SourceMap {
    "mappings": "AAAA,CAAC,MAAM,CAAC,IAAI,CAAC,CAAC,EAAE,CAAC,CCAjB;AACA;AAAA;AAAA,0DAA0E;ADC1E,CAAC,CAAC,MAAM;",
    "names": [],
    "sources": [
      "App.svelte",
      "",
    ],
    "version": 3,
  },
  "toString": [Function],
}
`;

exports[`preprocessor > should understand different formatting 1`] = `
{
  "code": "<script lang=\\"ts\\">
    import type 
    { 
        IconContextProps as IconProps 
    }	from 'phosphor-svelte';
  import IceCream from \\"phosphor-svelte/lib/IceCream\\";
  import FolderNotch from 'phosphor-svelte/lib/FolderNotch';
  import Flag
    from 'phosphor-svelte/lib/FlagCheckered';
  import Foo from \\"phosphor-svelte/lib/IceCream\\";
  import Gif from \\"phosphor-svelte/lib/Gif\\";
  import Spooky from \\"phosphor-svelte/lib/Ghost\\";
  import type { Hourglass } from \\"phosphor-svelte\\";
  import type { SvelteComponent, ComponentType } from 'svelte';
</script>
",
  "dependencies": [],
  "map": SourceMap {
    "mappings": "AAAA,CAAC,MAAM,CAAC,IAAI,CAAC,CAAC,EAAE,CAAC,CCAjB;AACA;AACA;AACA;AACA;AACA,sDAA6C;AAC7C;AACA;AACA;AACA,iDAC+C;AAC/C;AAAA;AAAA,mDAI2B;AAC3B;ADCA,CAAC,CAAC,MAAM;",
    "names": [],
    "sources": [
      "App.svelte",
      "",
    ],
    "version": 3,
  },
  "toString": [Function],
}
`;
